src/lib/operators/join_sort_merge/column_materializer.hpp

```diff
@@ -30,99 +32,82 @@ template <typename T>
- * the chunks. Each job then adds `samples_to_collect` samples to its passed
- * SampleRequest. All SampleRequests are later merged to gather a global sample
- * list with which the split values for the radix partitioning are determined.
- */
+using MaterializedSegmentList = std::vector<MaterializedSegment<T>>;
+
+// This data structure is passed as a reference to the jobs which materialize the chunks. Each job then adds
+// `samples_to_collect` samples to its passed SampleRequest. All SampleRequests are later merged to gather a global
+// sample list with which the split values for the radix partitioning are determined.
 template <typename T>
 struct Subsample {
-  explicit Subsample(ChunkOffset sample_count) : samples_to_collect(sample_count), samples() {}
+  explicit Subsample(ChunkOffset sample_count) : samples_to_collect(sample_count), samples(sample_count) {}
   const ChunkOffset samples_to_collect;
   std::vector<T> samples;
 };
 
-/**
- * Materializes a table for a specific segment and sorts it if required. Result is a triple of
- * materialized values, positions of NULL values, and a list of samples.
- **/
+// Materializes a column and sorts it if requested. Result is a triple of materialized values, positions of NULL
+// values, and a list of samples.
 template <typename T>
 class ColumnMaterializer {
  public:
   explicit ColumnMaterializer(bool sort, bool materialize_null) : _sort{sort}, _materialize_null{materialize_null} {}
 
  public:
-  /**
-   * Materializes and sorts all the chunks of an input table in parallel
-   * by creating multiple jobs that materialize chunks.
-   * Returns the materialized segments and a list of null row ids if materialize_null is enabled.
-   **/
-  std::tuple<std::unique_ptr<MaterializedSegmentList<T>>, std::unique_ptr<RowIDPosList>, std::vector<T>> materialize(
-      const std::shared_ptr<const Table> input, const ColumnID column_id) {
-    const ChunkOffset samples_per_chunk = 10;  // rather arbitrarily chosen number
+  // Materializes and sorts (if requested) all the chunks of an input table. For sufficiently large chunks, the
+  // materialization is parallelized. Returns the materialized segments and a list of null row ids if _materialize_null
+  // is true.
+  std::tuple<MaterializedSegmentList<T>, RowIDPosList, std::vector<T>> materialize(
+      const std::shared_ptr<const Table>& input, const ColumnID column_id) {
+    constexpr ChunkOffset SAMPLES_PER_CHUNK = 10;  // rather arbitrarily chosen number
     const auto chunk_count = input->chunk_count();
 
-    auto output = std::make_unique<MaterializedSegmentList<T>>(chunk_count);
-    auto null_rows = std::make_unique<RowIDPosList>();
+    auto output = MaterializedSegmentList<T>(chunk_count);
 
-    std::vector<Subsample<T>> subsamples;
+    auto null_rows_per_chunk = std::vector<RowIDPosList>(chunk_count);
+    auto subsamples = std::vector<Subsample<T>>{};
     subsamples.reserve(chunk_count);
```

Comment:
Would we know how to size it correctly? Is order important to this?

Reply:
I would be possible with one vector, but since we don't know the number of samples per chunk (might be a very small chunk), it would require a second loop. Due to the size of the samples, the `insert()` should not have a measurable effect.

Order does not matter.